internal/worker/objectstores3caller/manifold.go

```diff
@@ -0,0 +1,151 @@
+	if cfg.HTTPClientName == "" {
+		return errors.NotValidf("nil HTTPClientName")
+	}
+	if cfg.ServiceFactoryName == "" {
+		return errors.NotValidf("nil ServiceFactoryName")
+	}
+	if cfg.NewClient == nil {
+		return errors.NotValidf("nil NewClient")
+	}
+	if cfg.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if cfg.Clock == nil {
+		return errors.NotValidf("nil Clock")
+	}
+	return nil
+}
+
+// Manifold returns a manifold whose worker wraps an S3 Session.
+func Manifold(config ManifoldConfig) dependency.Manifold {
+	return dependency.Manifold{
+		Inputs: []string{
+			config.HTTPClientName,
+			config.ServiceFactoryName,
+		},
+		Output: outputFunc,
+		Start:  config.start,
+	}
+}
+
+// start returns a StartFunc that creates a S3 client based on the supplied
+// manifold config and wraps it in a worker.
+func (config ManifoldConfig) start(ctx context.Context, getter dependency.Getter) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	controllerConfigService, err := config.GetControllerConfigService(getter, config.ServiceFactoryName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	controllerConfig, err := controllerConfigService.ControllerConfig(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// If we're not using S3, then we don't need to start this worker.
+	if controllerConfig.ObjectStoreType() != objectstore.S3Backend {
+		return nil, dependency.ErrUninstall
+	}
+
+	var httpClient s3client.HTTPClient
+	if err := getter.Get(config.HTTPClientName, &httpClient); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return config.NewWorker(workerConfig{
+		ControllerService: controllerConfigService,
+		HTTPClient:        httpClient,
+		NewClient:         config.NewClient,
+		Logger:            config.Logger,
+		Clock:             config.Clock,
+	})
+}
+
+// outputFunc extracts a S3 client from a *s3caller.
+func outputFunc(in worker.Worker, out any) error {
+	inWorker, _ := in.(*s3Worker)
+	if inWorker == nil {
+		return errors.Errorf("in should be a %T; got %T", inWorker, in)
+	}
+
+	switch outPointer := out.(type) {
+	case *objectstore.Client:
+		*outPointer = inWorker
+	default:
+		return errors.Errorf("out should be *s3caller.Session; got %T", out)
+	}
+	return nil
+}
+
+// NewS3Client returns a new S3 client based on the supplied dependencies.
+func NewS3Client(url string, client s3client.HTTPClient, creds s3client.Credentials, logger s3client.Logger) (objectstore.Session, error) {
+	return s3client.NewS3Client(url, client, creds, logger)
+}
+
+// GetControllerConfigService is a helper function that gets a service from the
+// manifold.
+func GetControllerConfigService(getter dependency.Getter, name string) (ControllerService, error) {
+	return coredependency.GetDependencyByName[servicefactory.ControllerServiceFactory, ControllerService](getter, name, func(factory servicefactory.ControllerServiceFactory) ControllerService {
```

Comment:
You don't actually need the generic parameters here. They are implied by the signature.
The shortened line reads easier too.